### PR TITLE
[ActionsLog] Improve UX for loggable extensions checkboxes with a select all/none SEE #22699

### LINKED ISSUE #22627: [com_actionlogs] After deselecting all events to disable logging, checkboxes remain ticked
### Steps to reproduce the issue

Joomla 3.9 RC1 
Click Options in Action Logs screen
deselect all events to be logged (thus disabling logging) 
Click save

### Expected result

Page reloads with green configuration saved notice
Checkboxes are all NOT checked

### Actual result

Page reloads with green configuration saved notice
Checkboxes are all CHECKED again 

### System information (as much as possible)

Google Chrome, no JS errors in console

screen recording 2018-10-15 at 10 11 am

### PR DESCRIPTION
**Pull request by cyrez**
**EDIT**
## new PR here: #22699
 
 

Pull Request for Issue #22627.

### Summary of Changes
- Remove default attribute for `loggable_extensions` (Fix #22627 UX issue)
- Add a "Select All/None" for loggable extensions to help selecting all or clearing all on one click.

capture d ecran 2018-10-16 a 21 20 40

### Testing Instructions
- See issue #22627 
- Test actions logging options, all selected, some selected, none selected.

### CODE REVIEW

File: administrator/components/com_actionlogs/models/fields/logtype.php [hunk 1]
```diff
@@ -29,6 +29,84 @@ class JFormFieldLogType extends JFormFieldCheckboxes
 	 */
 	protected $type = 'LogType';
 
+	/**
+	 * Method to get the field input markup for check boxes.
+	 *
+	 * @return  string  The field input markup.
+	 *
+	 * @since   3.9.0
```
Comment (Quy) [inline administrator/components/com_actionlogs/models/fields/logtype.php]: Change to `__DEPLOY_VERSION__`

File: administrator/components/com_actionlogs/models/fields/logtype.php [hunk 2]
```diff
@@ -29,6 +29,84 @@ class JFormFieldLogType extends JFormFieldCheckboxes
...
+	 */
+	public function getInput()
+	{
+		$html = '';
```
Comment (Quy) [inline administrator/components/com_actionlogs/models/fields/logtype.php]: Remove and start/change line 62 `$html =`

File: administrator/components/com_actionlogs/models/fields/logtype.php [hunk 3]
```diff
@@ -29,6 +29,84 @@ class JFormFieldLogType extends JFormFieldCheckboxes
...
+				for (i = 0; i < events.length; i++) {
+					events[i].checked = value;
+				}
+			}
+		";
+
+		// Add the script to the document head
+		JFactory::getDocument()->addScriptDeclaration($script);
+
+		$html.= '<div class="well well-small">';
```
Comment (Quy) [inline administrator/components/com_actionlogs/models/fields/logtype.php]: Add space before `.=`

**Review by mbabker — APPROVED** [review]
Looks fine on review, requires tests

File: administrator/components/com_actionlogs/models/fields/logtype.php (hunk 2 above)
Comment (cyrez) [reply administrator/components/com_actionlogs/models/fields/logtype.php]: Thanks @Quy !

I did the changes.

I thought when use concatenation operators, the first one should be empty (to help to keep code diffs minimal)...
Was it changed as a rule or i was wrong ? 

Note: some other core files still use first empty (PR needed for code standards ?).
eg https://github.com/joomla/joomla-cms/blob/staging/libraries/vendor/joomla/uri/src/AbstractUri.php#L150
```
		$uri = '';
		$uri .= $parts & static::SCHEME ? (!empty($this->scheme) ? $this->scheme . '://' : '') : '';
		$uri .= $parts & static::USER ? $this->user : '';
		$uri .= $parts & static::PASS ? (!empty($this->pass) ? ':' : '') . $this->pass . (!empty($this->user) ? '@' : '') : '';
		$uri .= $parts & static::HOST ? $this->host : '';
		$uri .= $parts & static::PORT ? (!empty($this->port) ? ':' : '') . $this->port : '';
		$uri .= $parts & static::PATH ? $this->path : '';
		$uri .= $parts & static::QUERY ? (!empty($query) ? '?' . $query : '') : '';
		$uri .= $parts & static::FRAGMENT ? (!empty($this->fragment) ? '#' . $this->fragment : '') : '';

```

File: administrator/components/com_actionlogs/models/fields/logtype.php [hunk 4]
```diff
@@ -29,6 +29,82 @@ class JFormFieldLogType extends JFormFieldCheckboxes
...
+
+		foreach ($options as $i => $option)
+		{
+			// Initialize some option attributes.
+			$checked = in_array((string) $option->value, $checkedOptions, true) ? 'checked' : '';
+			$oid     = $this->id . $i;
+			$value   = htmlspecialchars($option->value, ENT_COMPAT, 'UTF-8');
+
+			$html .= '<label for="' . $oid . '" class="checkbox">';
+			$html .= sprintf($format, $oid, $this->name, $value, ' ' . $checked);
```
Comment (Quy) [inline administrator/components/com_actionlogs/models/fields/logtype.php]: `' ' . ` can be removed since the space is already accounted in `$format`

Comment (cyrez) [reply administrator/components/com_actionlogs/models/fields/logtype.php]: Well found @Quy !

Removed extra space.

**Comment by Quy** [comment]
It is still an issue when editing a profile.

Edit a Super User profile
Go to `User Actions Log Options`
Select `None`
Click `Save`

All checkboxes are checked.

**Comment by cyrez** [comment]
@Quy Oh! I didn't even see there was this field in super user edit...

To be honnest, i don't see yet how the core extensions are set by default, as no default attribute in plugin system actionlogs form...

**Comment by cyrez** [comment]
@Quy Did you test super user with latest staging including https://github.com/joomla/joomla-cms/pull/21983 ?

**Comment by Quy** [comment]
Yes tested with latest staging including #21983.

**Comment by SharkyKZ** [comment]
The plugin will need to be adjusted. But the field also needs to remove `checked` attribute from its options.

Test case:
Save `com_actionlogs` configuration with some or all events selected.
Deselect all events and save again.
No events are saved but all events appear checked.

**Comment by cyrez** [comment]
Thanks @SharkyKZ 
Do you have tested this PR ?

> Test case:
Save com_actionlogs configuration with some or all events selected.
Deselect all events and save again.
No events are saved but all events appear checked.

With this PR, it works for me with `com_actionlogs` configuration as i've removed there the default attribute from `config.xml`.
But in plugin system actionlogs, no default attribute for user edit form `actionlogs.xml`
So it should act as for global options as same form field used.

The issue is with this PR not really the same as it was before.

Maybe in menu/plugin, the issue is the save function does not allow empty submitted data, and then replace by previously saved data.

@SharkyKZ your help could be appreciated to know if a change in form data processed to be done in the plugin, or if another issue elsewhere...

**Comment by cyrez** [comment]
@SharkyKZ I've found why this PR does not have effect on super user edit!

In plugin, it is forced to save only if isset. But if empty, it won't update `extensions` column, and so keep previous setting.

**Fix**
Adding the following `else` statement for this `if` statement here: https://github.com/joomla/joomla-cms/pull/21983/files#diff-bbf838ec8885d4ea43274f7b94e17808R343:
```
else
{
	$values[] = $this->db->quoteName('extensions') . ' = ""';
}
```
Would fix issue.

Note: i can add this change to this PR if you think it's ok with this ?

**Comment by SharkyKZ** [comment]
#21983 was written to maintain the existing behavior of form fields. See https://github.com/joomla/joomla-cms/pull/21983#issuecomment-423062571.

Your fix seems OK. But maybe we should store empty JSON array (`[]`) instead of an empty string. IDK about this.

Also the field itself needs updating to remove `checked` attribute by default. Otherwise, after deselecting all options and saving they appear checked again, even when `default` value is not set. See this part https://github.com/joomla/joomla-cms/blob/a4479d39b71c574a793b2bfda06a0332caeb130f/administrator/components/com_actionlogs/models/fields/logtype.php#L125

**Comment by mbabker** [comment]
> Your fix seems OK. But maybe we should store empty JSON array (`[]`) instead of an empty string. IDK about this.

If a column or data source is expecting a JSON string, then it should always be stored as something that doesn't cause `json_decode()` to set an error flag.  We've got too many places in core where this is being hacked around already.

**Comment by cyrez** [comment]
Will redo this PR with changes needed (i did something wrong with udpating this one, and better to get things clear) ;-)

**Comment by cyrez** [comment]
PR redo here: https://github.com/joomla/joomla-cms/pull/22699
Thanks for comments and help here! 👍

**Pull request closed without merging**